### DISCUSSION TITLE
ErrorExceptions because of Incompatible return types with SplFileInfo

### DISCUSSION THREAD

**Discussion by ralphjsmit** [livewire/livewire#4530] [7. Bugs (archived)] [answered]
Hey Caleb and Josh, 

Thanks for maintaining and steadily improving this absolutely amazing tool!

I'd like to report an error / notice that I've been getting on two or three production environments. All environments are running on PHP 8.1, so I guess that it's related to PHP 8.1.

In short, the full error is this:

*Return type of Livewire\TemporaryUploadedFile::getRealPath() should either be compatible with SplFileInfo::getRealPath(): string|false, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice.*

The `Livewire\TemporaryUploadedFile::getRealPath()` looks like this:

```php
public function getRealPath()
{
    //
}
```

I took a little look at the `Livewire\TemporaryUploadedFile` and when you dig through all the extends, you can see that the `TemporaryUploadedFile` class extends from the `SplFileInfo` class. My PhpStorm stub shows that the function `SplFileInfo::getRealPath()` looks like this:

```php
public function getRealPath(): string|false {}
```

As you can see, the return type doesn't match, which is what's throwing the error. 

There are also many other similar errors:

*Return type of Livewire\TemporaryUploadedFile::getPath() should either be compatible with SplFileInfo::getPath(): string, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice*

<img width="990" alt="Schermafbeelding 2022-01-17 om 11 37 15" src="https://user-images.githubusercontent.com/59207045/149754281-83afc40e-eb28-4786-b350-a08673973de1.png">

I'm not fully sure, but to me it looks like we could add the correct return types to all Livewire functions and fix the problem with that. PHP supports return types since PHP 7, so I guess that it's safe now to use them?

I don't have much knowledge of the internals of Livewire, but if I can be of any help, please let me know!

Thanks!

**Comment by joshhanley** [comment, marked as answer]
@ralphjsmit thanks for being willing to dig in on this! There is currently an open PR #4485 which should resolve this issue 🙂

**Reply by ralphjsmit** [reply]
Awesome, thanks! I see in the PR that you/Caleb aren't considering updating the methods to be compatible with the updated `SplFileInfo`. Or is that more for a potential Livewire `v3` in the future?

**Reply by joshhanley** [reply]
@ralphjsmit yeah we originally weren't, but I think we may need to, for compatibility with Laravel 9 as it has a dependency on Symphony 6 which requires those method signatures.
